framework/src/outputs/Checkpoint.C

```diff
@@ -75,60 +87,80 @@ Checkpoint::directory() const
 
-  // Write out the restartable mesh meta data if there is any (only on processor zero)
-  if (processor_id() == 0)
-  {
-    for (auto & map_pair :
-         libMesh::as_range(_app.getRestartableDataMapBegin(), _app.getRestartableDataMapEnd()))
-    {
-      const RestartableDataMap & meta_data = map_pair.second.first;
-      const std::string & suffix = map_pair.second.second;
-      const std::string filename(curr_file_struct.checkpoint + "/meta_data" + suffix +
-                                 _restartable_data_io.getRestartableDataExt());
+    curr_file_struct.checkpoint = current_file + getMeshFileSuffix(_binary);
+    curr_file_struct.system = current_file + _restartable_data_io.getESFileExtension(_binary);
+    curr_file_struct.restart = current_file + _restartable_data_io.getRestartableDataExt();
 
-      curr_file_struct.restart_meta_data.emplace(filename);
-      _restartable_data_io.writeRestartableData(filename, meta_data);
+    // Write the checkpoint file
+    io.write(curr_file_struct.checkpoint);
+
+    // Write out the restartable mesh meta data if there is any (only on processor zero)
+    if (processor_id() == 0)
+    {
+      for (auto & map_pair :
+           libMesh::as_range(_app.getRestartableDataMapBegin(), _app.getRestartableDataMapEnd()))
+      {
+        const RestartableDataMap & meta_data = map_pair.second.first;
+        const std::string & suffix = map_pair.second.second;
+        const std::string filename(curr_file_struct.checkpoint + "/meta_data" + suffix +
+                                   _restartable_data_io.getRestartableDataExt());
+
+        curr_file_struct.restart_meta_data.emplace(filename);
+        _restartable_data_io.writeRestartableData(filename, meta_data);
+      }
     }
-  }
 
-  // Write the system data, using ENCODE vs WRITE based on ascii vs binary format
-  _es_ptr->write(curr_file_struct.system,
-                 EquationSystems::WRITE_DATA | EquationSystems::WRITE_ADDITIONAL_DATA |
-                     EquationSystems::WRITE_PARALLEL_FILES,
-                 renumber);
+    // Write the system data, using ENCODE vs WRITE based on ascii vs binary format
+    _es_ptr->write(curr_file_struct.system,
+                   EquationSystems::WRITE_DATA | EquationSystems::WRITE_ADDITIONAL_DATA |
+                       EquationSystems::WRITE_PARALLEL_FILES,
+                   renumber);
+
+    // Write the restartable data
+    _restartable_data_io.writeRestartableDataPerProc(curr_file_struct.restart, _restartable_data);
 
-  // Write the restartable data
-  _restartable_data_io.writeRestartableDataPerProc(curr_file_struct.restart, _restartable_data);
+    // Remove old checkpoint files
+    updateCheckpointFiles(curr_file_struct);
 
-  // Remove old checkpoint files
-  updateCheckpointFiles(curr_file_struct);
+    // Stop outputting the checkpoint if this is a signaled/autosave checkpoint
+    if (Moose::autosave_flag)
+    {
+      _should_output = false;
```

Comment:
This might not be right... Are we reusing the Checkpoint instance if it's already there or not? If we reuse it, then _should_output could be independently true based on the normal Checkpoint output triggers. We need to be purposeful in handling the multiple instances case.